src/librustc_mir/dataflow/impls/reaching_defs.rs

```diff
@@ -0,0 +1,414 @@
+    /// `mir::Body` is entered (e.g. the value of an argument).
+    pub location: Option<Location>,
+}
+
+impl Definition {
+    fn is_indirect(&self) -> bool {
+        self.kind == DefKind::Indirect
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum DefKind<T> {
+    /// An entire variable is defined directly, e.g. `_1 = ...`.
+    DirectWhole(T),
+
+    /// Some part of a variable (a field or an array offset) is defined directly, e.g. `_1.field =
+    /// ...` or `_1[2] = ...`.
+    DirectPart(T),
+
+    /// An unknown variable is defined through a pointer, e.g. `(*_1) = ...`.
+    ///
+    /// In general, we cannot know where `_1` is pointing: it may reference any local in the
+    /// current `Body` (whose address has been observed) or somewhere else entirely (e.g. a static,
+    /// the heap, or a local further down the stack).
+    Indirect,
+}
+
+impl<T> DefKind<T> {
+    /// Returns `Some(T)` if the `DefKind` is direct or `None` if it is indirect.
+    pub fn direct(self) -> Option<T> {
+        match self {
+            DefKind::DirectWhole(x) | DefKind::DirectPart(x) => Some(x),
+            DefKind::Indirect => None,
+        }
+    }
+
+    /// Converts a `DefKind<T>` to a `DefKind<U>` by applying a function to the value contained in
+    /// either `DirectWhole` or `DirectPart`.
+    fn map<U>(self, f: impl FnOnce(T) -> U) -> DefKind<U> {
+        match self {
+            DefKind::DirectWhole(x) => DefKind::DirectWhole(f(x)),
+            DefKind::DirectPart(x) => DefKind::DirectPart(f(x)),
+            DefKind::Indirect => DefKind::Indirect,
+        }
+    }
+}
+
+impl<T> DefKind<Option<T>> {
+    /// Converts a `DefKind` of an `Option` into an `Option` of a `DefKind`.
+    ///
+    /// The result is `None` if either direct variant contains `None`. Otherwise, the result is
+    /// `Some(self.map(Option::unwrap))`.
+    fn transpose(self) -> Option<DefKind<T>> {
+        match self {
+            DefKind::DirectWhole(Some(x)) => Some(DefKind::DirectWhole(x)),
+            DefKind::DirectPart(Some(x)) => Some(DefKind::DirectPart(x)),
+            DefKind::Indirect => Some(DefKind::Indirect),
+
+            DefKind::DirectWhole(None) | DefKind::DirectPart(None) => None,
+        }
+    }
+}
+
+impl<'a, 'tcx> DefKind<&'a mir::PlaceBase<'tcx>> {
+    fn of_place(place: &'a mir::Place<'tcx>) -> Self {
+        place.iterate(|base, projections| {
+            let mut is_whole = true;
+            for proj in projections {
+                is_whole = false;
+                if proj.elem == mir::ProjectionElem::Deref {
+                    return DefKind::Indirect;
+                }
+            }
+
+            if is_whole {
+                DefKind::DirectWhole(base)
+            } else {
+                DefKind::DirectPart(base)
+            }
+        })
+    }
+}
+
+pub struct DefVisitor<F>(pub F);
+
+impl<'tcx, F> Visitor<'tcx> for DefVisitor<F>
```

Comment:
I'm worried about the correctness and efficiency of `DefVisitor` and `UseVisitor`. Is this a good approach? For `DefVisitor` at least, we only have to visit the left-hand side of any `StatementKind::Assign`.